File: flutter-device-binding/android/app/src/main/kotlin/com/doa/example/devicebinding/flutter/HealthIdChannel.kt
package com.doa.example.devicebinding.flutter

import android.content.Intent
import android.net.Uri
import android.util.Base64
import android.content.pm.ApplicationInfo
import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.security.MessageDigest
import java.security.SecureRandom

/** Browser/app handoff only. OAuth exchange stays in Dart; no tokens are logged. */
class HealthIdChannel(private val activity: MainActivity) {
    private companion object {
        const val TAG = "health-id"
    }

    /** Release builds say nothing at all; this reads the manifest flag, not a
     *  build-time constant, so no extra Gradle feature has to be switched on. */
    private val debug: Boolean
        get() = activity.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0

    private var pending: MethodChannel.Result? = null
    private var redirect: Uri? = null
    fun register(messenger: BinaryMessenger) {
        MethodChannel(messenger, "com.doa.example.devicebinding.flutter/health_id")
            .setMethodCallHandler { call, result ->
                when (call.method) {
                    "createProof" -> {
                        fun random(): String {
                            val bytes = ByteArray(32)
                            SecureRandom().nextBytes(bytes)
                            return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
                        }
                        val verifier = random()
                        val challenge = Base64.encodeToString(MessageDigest.getInstance("SHA-256")
                            .digest(verifier.toByteArray(Charsets.US_ASCII)),
                            Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
                        result.success(mapOf("verifier" to verifier, "challenge" to challenge,
                            "state" to random(), "nonce" to random()))
                    }
                    "authorize" -> {
                        if (pending != null) {
                            result.error("BUSY", "Authentication is already running", null)
                        } else {
                            try {
                                val target = Uri.parse(call.argument<String>("url")!!)
                                val callback = Uri.parse(call.argument<String>("redirect")!!)
                                require(callback.scheme == "https" && !callback.host.isNullOrEmpty())
                                require(target.scheme !in listOf("http", "file", "intent", "javascript", "content"))
                                pending = result
                                redirect = callback
                                activity.startActivity(Intent(Intent.ACTION_VIEW, target)
                                    .addCategory(Intent.CATEGORY_BROWSABLE))
                            } catch (_: Exception) {
                                pending = null
                                redirect = null
                                result.error("UNAVAILABLE", "Could not open authenticator", null)
                            }
                        }
                    }
                    "cancel" -> { cancel(); result.success(null) }
                    else -> result.notImplemented()
                }
            }
    }
    fun handle(intent: Intent) {
        val uri = intent.data ?: return
        // Scheme, host, port and path only - never the query, which carries the
        // code and state. Debug builds only.
        if (debug) Log.d(TAG, "callback for ${uri.scheme}://${uri.host}${uri.path}")
        val expected = redirect ?: return
        if (uri.scheme != expected.scheme || uri.host != expected.host ||
            uri.port != expected.port || uri.path != expected.path) {
            if (debug) Log.d(TAG, "callback ignored, not the one awaited")
            return
        }
        val result = pending
        pending = null
        redirect = null
        result?.success(uri.toString())
    }
    fun cancel() {
        val result = pending
        pending = null
        redirect = null
        result?.error("CANCELLED", "Authentication cancelled", null)
    }
}

File: flutter-device-binding/ios/Runner/Channels/HealthIdChannel.swift
import CryptoKit
import Flutter
import Foundation
import UIKit

/// The Health-ID hand-off, iOS side. Mirrors `HealthIdChannel.kt`.
///
/// The same shape as Android on purpose: the proof is made here so the verifier
/// never crosses into Dart before it is needed, the insurer's page opens in the
/// system browser rather than an in-app web view, and the call stays pending
/// until the universal link comes back. The callback is matched on scheme,
/// host, port and path against the registered redirect, and anything else is
/// ignored rather than answered - a foreign link must not complete someone's
/// sign-in.
///
/// Safari is used rather than `ASWebAuthenticationSession` because the redirect
/// is an https universal link, which that API can only capture from iOS 17.4.
/// Opening the browser and receiving the link through the app is what Android
/// does and works on every supported version.
final class HealthIdChannel {
  static let name = "com.doa.example.devicebinding.flutter/health_id"

  private var pending: FlutterResult?
  private var expectedRedirect: URLComponents?

  func register(with messenger: FlutterBinaryMessenger) -> FlutterMethodChannel {
    let channel = FlutterMethodChannel(name: Self.name, binaryMessenger: messenger)
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result)
    }
    return channel
  }

  private func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    switch call.method {
    case "createProof":
      let verifier = Self.random()
      let challenge = Data(SHA256.hash(data: Data(verifier.utf8))).base64URLEncoded
      result([
        "verifier": verifier,
        "challenge": challenge,
        "state": Self.random(),
        "nonce": Self.random(),
      ])

    case "authorize":
      guard pending == nil else {
        result(FlutterError(
          code: "BUSY", message: "Authentication is already running", details: nil))
        return
      }
      guard let raw = call.optionalString("url"),
            let url = URL(string: raw),
            let redirect = call.optionalString("redirect"),
            let expected = URLComponents(string: redirect)
      else {
        result(FlutterError(
          code: "UNAVAILABLE", message: "Could not open authenticator", details: nil))
        return
      }
      pending = result
      expectedRedirect = expected
      UIApplication.shared.open(url, options: [:]) { [weak self] opened in
        guard let self, !opened else { return }
        self.finish(FlutterError(
          code: "UNAVAILABLE", message: "Could not open authenticator", details: nil))
      }

    case "cancel":
      finish(FlutterError(
        code: "CANCELLED", message: "Authentication cancelled", details: nil))
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Called by the app delegate for every universal link. Returns true only
  /// when this was the callback being waited for, so the rest of the app can
  /// go on handling its own links.
  @discardableResult
  func handle(callback url: URL) -> Bool {
    guard pending != nil,
          let expected = expectedRedirect,
          let incoming = URLComponents(url: url, resolvingAgainstBaseURL: false),
          incoming.scheme == expected.scheme,
          incoming.host == expected.host,
          incoming.port == expected.port,
          incoming.path == expected.path
    else { return false }
    finish(url.absoluteString)
    return true
  }

  /// A process death or a dismissed browser leaves nothing pending to resume;
  /// the Dart side times the flow out and starts the next one with fresh proof.
  private func finish(_ value: Any?) {
    let result = pending
    pending = nil
    expectedRedirect = nil
    result?(value)
  }

  /// 32 bytes from the system CSPRNG, base64url without padding: a PKCE
  /// verifier, and the same source for `state` and `nonce`.
  private static func random() -> String {
    var bytes = [UInt8](repeating: 0, count: 32)
    let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    precondition(status == errSecSuccess, "The system CSPRNG failed")
    return Data(bytes).base64URLEncoded
  }
}
